tripleblindmarket/private-kit
# Feature/tb jts whitelabel

This pr is a handful of commits that work towards introducing a white labeling strategy to build the application for either location based or Bluetooth based contact tracing strategies.  Given that one app cannot have both bluetooth and location based data for privacy concerns we need to be able to build the app binary with different permission settings in info.plist, in order to do this we introduce two build targets in xcode, GPS and BTE, that have different info.plist and bundle ids, but share all of the other native libraries and code. We exposes the tracing strategy that the given build is using to the js side over the bridge via a 'getTracingStrategy' function.

Ive left the commits separate for ease of review and i would recommend reviewing one at a time.

Note that this work only sets up the distinction between the gps and bte based builds. It does not implement any bluetooth permission work, it will just give us some space to do so.


### Running GPS

```
yarn run-ios-gps
```

![gps](https://user-images.githubusercontent.com/16049495/82203605-8b3f7500-98d1-11ea-8c57-619e1e26affe.gif)


### Running BTE

```
yarn run-ios-bte
```

![bte](https://user-images.githubusercontent.com/16049495/82203652-94c8dd00-98d1-11ea-8231-12ab4ff8980d.gif)


## Commit 1:

Use modules imports for assets …
41fec15
Why:
Currently we are importing png and svg assets individually from
shared/assets. This leads to a fair amount of churn in the import
statements and is somewhat of a pain to maintain overtime. By bundling
the assets in a common module we can reduce a bit of cognitive overhead
by allowing designers and developers to always be importing assets from
same place, leading to some significant speed gains during iterations on
design implementations. Additionally,this will also allow us a place to
put some type safety if we ever introduce typescript to the project.
There is no technical gains to be realized with this strategy, it is
purely to make life a bit easier for designers.

This commit:
Introduces an /shared/assets/index.js to import icons and images from.
No new functionality has been added.

----

This commit is perhaps not necessary, but while getting familiar with the app I ended up using this clean up as a way to build some context, I was also concerned about missing some path updates when moving the assets folder out of shared in the following commit and this refactor made it less likely for me to make a mistake.

## Commit 2

Introduce gps and bte ios targets …
67ade64
Why:
We would like to introduce bluetooth to the application for bluetooth
based contact tracing. Given the privacy concerns of having both
location and bluetooth data available to the same application, google
and apple do not allow for both location and bluetooth to be available
at the same time.

Hence in order to service both location and bluetooth
we will need to provide separate apps with separate info.plists and
bundle identifiers. At the same time we would like to share a much
application logic between the two applications as to reduce the
maintenance and organizational overhead. Ideally as close to 100% of the
javascript layer is common and for the ios layer we only need to change
the info.plists.

This commit:

Introduces two build targets, GPS and BTE, to the xcode project which
are setup to read from the corresponding groups for config and info.plist
A COVIDSafePathsConfig file was introduced to expose a
'getTracingStrategy' function to the react native side so that the js
runtime can determine which tracing strategy the bundle will be using.
This will allow us flexibility to programmatically determine which
tracing strategy is being used either within a given react component or
to conditional render components altogether. This also allows use to
maximize the amount of shared native code.

To build the corresponding targets via the react native cli, you can
pass the --scheme flag with the corresponding target:

```
react-native run-ios --scheme "GPS"
react-native run-ios --scheme "BTE"
```

We updated the bundle ids to `org.pathcheck.covid-safepaths.gps` and
`org.pathcheck.covid-safepaths.bte` respectively. Depending on app store
considerations, it might be easier to revert
`org.pathcheck.covid-safepaths.gps` to `org.pathcheck.covid-safepaths`

Note that theres a fair amount of location information storing logic
currently in AppDelegate, a future commit might want to refactor this so
that the gps specific code is only built in the GPS target builds.

## Commit 3

ollback js code spliting …
d61cf20
Why:
Given that we are using the two build targets of GPS and BTE on the
native side, it is no longer necessary to split the code on the
javascript side.

This commit:
Moves all of the application logic from gps/app back to app/
Additionally assets where moved out of the shared folder into app/ so
that we could remove the shared/ directory altogether. This commit is a
refactor and introduces no new functionality.

## Commit 4

Refactor onboarding5 to use functional component …
4b982d0
Why:
We would like to introduce some information to the component via a
context. It is best practice to use hooks and so we would like to use
the useContext hook for this. Currently the onboarding5 component is
build with the class syntax keeping us from doing this.

This commit:
Refactors the Onbarding5 component to use the functional style. No new
functionality has been added.

## Commit 5

Introduce covid safe paths config …
3c90fba
Why:
We would like to be able to conditionally render react components based
off of the tracing strategy being used by the app. Given that the app
can be built with multiple ios targets we need to fetch this information
from over the bridge.

This commit:
Introduces the COVIDSafePathsConfig to the app, which uses the
NativeModules api to fetch the tracing strategy and stores it in a
config object to be used as needed in components.

Note that this commit only introduces the pattern to a handful of
existing components. It does not exhaustively update every spot that
needs to render conditionally based off of the tracing strategy.

At first we considered using a context for this, we rolled that back
since it seemed to be a bit overkill at this point. It might make sense
to reintroduce in a future commit if there is significant application
logic that needs to be run and shared amongst many components related to
this.

## Commit 5

Update readme, package.json for bte/gps targets …
5ccfb95
This commit updates the readme and package.json to include the info for
building to either the gps or bte target for ios


## How to test:

Run the apps locally with the two schemes.
```
yarn run-ios-gps
```

```
yarn run-ios-bte
```


Remaining work to be done:

- do a similar set of work for the android side

- verify that this setup still works with the fastlane setup and if not adjusting to do so. 
That is, I wasnt able to verify that the production builds are still working, for sake of time. It might makes sense it introduce build targets for development and staging for each of the flavors of the app.

- set up translations for bte / gps
This one could probably use some thought. Something like : t('label.my_text', 'bte') or t('label.my_text', 'gps') might makes sense. We potentially have a combinatoric problem of two app flavors for every language and setting up the translations in a maintainable way will be important.

- consider moving the swift location storage code out of AppDelegate and into the gps group so that it is only included in the gps target build.
